### PR TITLE
IC-1718 Action Plan Approval Notification

### PR DESCRIPTION
**Pull request by mahalrmoj**
## What does this pull request do?
This PR notifies Delius of an approved action plan (via a call to community-api).

## What is the intent behind these changes?
The intention of this PR falls under the general intention to notify/update Delius of all significant events in planning and delivering a referral to a service user.

### CODE REVIEW

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt [hunk 1]
```diff
@@ -25,6 +26,9 @@ interface CommunityAPIService {
     val primeProviderName = referral.intervention.dynamicFrameworkContract.primeProvider.name
     return "$description for $contractTypeName Referral ${referral.referenceNumber} with Prime Provider $primeProviderName\n$url"
   }
+
+  fun camelCase(word: String) =
+    word.lowercase().replaceFirstChar { if (it.isLowerCase()) it.titlecase(Locale.getDefault()) else it.toString() }
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:31]: this appears to just capitalize the first letter of the input string? what's it supposed to be doing?

Comment (andrewburrowsmoj) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:31]: should be able to just do word.capitalize()

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt [hunk 2]
```diff
@@ -151,7 +155,8 @@ class CommunityAPIEndOfServiceReportEventService(
 @Service
 class CommunityAPIActionPlanEventService(
   @Value("\${interventions-ui.baseurl}") private val interventionsUIBaseURL: String,
-  @Value("\${interventions-ui.locations.submit-action-plan}") private val interventionsUIActionPlanLocation: String,
+  @Value("\${interventions-ui.locations.submit-action-plan}") private val interventionsUISubmittedActionPlanLocation: String,
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:158]: won't these URLs be the same? /probation-practitioner/action-plan/{id} ?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt [hunk 3]
```diff
@@ -160,28 +165,36 @@ class CommunityAPIActionPlanEventService(
...
+        postNotificationRequest(event, url, event.actionPlan.approvedAt!!)
       }
     }
   }
 
-  private fun postNotificationRequest(event: ActionPlanEvent, url: String) {
+  private fun postNotificationRequest(event: ActionPlanEvent, url: String, eventTime: OffsetDateTime) {
 
     val referral = event.actionPlan.referral
+    val status = camelCase(event.type.name)
```
Comment (tomsmyers) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:190]: i would prefer an explicit `when` block here. i think it's clearer what it's doing, and it decouples the event type enum name from the delius status.

```
val status = when(event.type) {
  APPROVED -> "Approved"
  SUBMITTED -> "Submitted"
}

File: src/main/resources/application.yml [hunk 4]
```diff
@@ -88,6 +88,7 @@ interventions-ui:
     sent-referral: "/service-provider/referrals/{id}/details"
     cancelled-referral: "/service-provider/referrals/{id}/details"
     submit-action-plan: "/service-provider/referrals/{id}/progress" # fixme
+    approved-action-plan: "/service-provider/referrals/{id}/progress" # fixme
```
Comment (tomsmyers) [inline src/main/resources/application.yml:91]: i think we should clarify these urls. for urls stored in delius they must be PP facing URLs. 

for email notifications to SPs they must be SP facing URLs.

etc etc

**Review by tomsmyers — COMMENTED** [review]
couple of small things

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt (hunk 1 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:31]: will make explicit as per your later comment

val status = when(event.type) {
  APPROVED -> "Approved"
  SUBMITTED -> "Submitted"
}

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt (hunk 2 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:158]: Can consolidate into one but will there ever be the need to make them different?

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt (hunk 3 above)
Comment (mahalrmoj) [reply src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIService.kt:190]: Will do

File: src/main/resources/application.yml (hunk 4 above)
Comment (mahalrmoj) [reply src/main/resources/application.yml:91]: Good point, will make these probation-practitioner

**Review by tomsmyers — APPROVED** [review]

File: src/main/resources/application.yml [hunk 5]
```diff
@@ -84,14 +84,20 @@ notify:
     end-of-service-report-submitted: "f79cb53f-dc5c-42aa-947e-f599f9ad7942"
 
 interventions-ui:
+  # links added to email notifications
   locations:
```
Comment (sldblog) [inline src/main/resources/application.yml:88]: Not sure links are naturally categorised to "email links" and "delius links". There will be overlap

Can we use `probation-links` and `provider-links` as groups? E.g. `interventions-ui.probation-links.session-feedback` and `interventions-ui.provider-links.submit-action-plan`

File: src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ActionPlanService.kt [hunk 6]
```diff
@@ -91,7 +91,7 @@ class ActionPlanService(
 
     actionPlanSessionsService.createUnscheduledSessionsForActionPlan(actionPlan)
     val approvedActionPlan = actionPlanRepository.save(actionPlan)
-    actionPlanEventPublisher.actionPlanSubmitEvent(approvedActionPlan)
```
Comment (sldblog) [inline src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/ActionPlanService.kt:94]: _nit_ Would be nice for this to be in past tense too

File: src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIActionPlanEventServiceTest.kt [hunk 7]
```diff
@@ -50,6 +53,25 @@ class CommunityAPIActionPlanEventServiceTest {
...
+    communityAPIService.onApplicationEvent(event)
+
+    verify(communityAPIClient).makeAsyncPostRequest(
+      "/secure/offenders/crn/X123456/sentence/1234/notifications/context/commissioned-rehabilitation-services",
+      NotificationCreateRequestDTO(
+        "ACC",
+        sentAtDefault,
+        event.actionPlan.referral.id,
+        approvedAtDefault,
+        "Action Plan Approved for Accommodation Referral XX1234 with Prime Provider Harmony Living\n" +
```
Comment (sldblog) [inline src/test/kotlin/uk/gov/justice/digital/hmpps/hmppsinterventionsservice/service/CommunityAPIActionPlanEventServiceTest.kt:69]: _nit_
From what I've seen, content is usually preferred to not have camel case, eg.
```suggestion
        "Action plan approved for Accommodation referral XX1234 with prime provider Harmony Living\n" +
```

**Review by sldblog — COMMENTED** [review]
Nice! 👍 

Two nits, one question

**Review by sldblog — APPROVED** [review]

**Pull request merged**